File: src/main/java/features/dicomimage/data/DicomImage.java
package features.dicomimage.data;


import core.data.AttributeCollection;
import core.data.medimage.MedImage;

import java.awt.image.BufferedImage;

/**
 * Dummy implementation of dicom image
 */
//TODO: (avlomakin) move to core.data
public class DicomImage extends MedImage {

    private BufferedImage pixelData;

    public DicomImage(BufferedImage pixelData) {
        this.pixelData = pixelData;
    }

    private DicomImage(AttributeCollection attributes, BufferedImage pixelData)
    {
        super(attributes);
        this.pixelData = pixelData;
    }

    @Override
    public BufferedImage getImage() {
        return pixelData;
    }

    @Override
    public void setImage(BufferedImage pixels) {
        pixelData = pixels;
    }

    @Override
    public DicomImage clone() {
        return new DicomImage(getAttributes().clone(), pixelData);
    }

    @Override
    public String getExtension() {
        return "DICOM";
    }
}

File: src/main/java/features/dicomimage/data/DicomImage.kt
package features.dicomimage.data

import com.pixelmed.dicom.AttributeList
import core.data.medimage.MedImage
import java.awt.image.BufferedImage

/**
 * Dummy implementation of dicom image
 */
//TODO: (avlomakin) move to core.data
class DicomImage : MedImage {

    override val attributes : DicomAttributeCollection

    private var _image : BufferedImage? = null
    private var _imageAttributesVersion = -1
    override val image: BufferedImage
        get(){
            if(_imageAttributesVersion != attributes.pixelDataAttributesVersion){
                log.info("attributes has been updated, recreating image")
                _image = attributes.buildHumanReadableImage()
                _imageAttributesVersion = attributes.pixelDataAttributesVersion
            }

            if(_image == null){
                log.warn("no image is presented, but versions are equal. Forcing image generation")
                _image = attributes.buildHumanReadableImage()
            }

            if(_image == null)
                throw MirfDicomException("failed to generate human readableImage")

            return _image!!
        }

    override val extension: String
        get() = "DICOM"

    constructor(attributes: DicomAttributeCollection) : super() {
        this.attributes = attributes
    }

    override fun clone(): DicomImage {
        return DicomImage(attributes.clone())
    }


    companion object {
        /**
         * Creates dicom image with no attributes
         */
        fun createEmpty(): DicomImage{
            return DicomImage(DicomAttributeCollection(AttributeList()))
        }
    }
}
